tests/integration/order/test_creator.py

```diff
@@ -218,3 +224,96 @@ def test_multi_sites(self):
                     site=site2)
         order2 = Order.objects.get(number='12346')
         self.assertEqual(order2.site, site2)
+
+
+class TestPlaceOrderWithVoucher(TestCase):
+
+    def test_single_usage(self):
+        user = AnonymousUser()
+        basket = factories.create_basket()
+        creator = OrderCreator()
+
+        voucher = factories.VoucherFactory(usage=Voucher.SINGLE_USE)
+        voucher.offers.add(factories.create_offer(offer_type='Voucher'))
+        basket.vouchers.add(voucher)
+
+        place_order(creator, basket=basket, order_number='12346', user=user)
+        assert voucher.applications.count() == 1
+
+        # Make sure the voucher usage is rechecked
+        with pytest.raises(ValueError):
+            place_order(creator, basket=basket, order_number='12347', user=user)
+
+
+class TestConcurrentOrderPlacement(TransactionTestCase):
+
+    def test_single_usage(self):
+        user = AnonymousUser()
+        creator = OrderCreator()
+        product = factories.ProductFactory(stockrecords__num_in_stock=1000)
+
+        # Make the order creator a bit more slow too reliable trigger
+        # concurrency issues
+        org_create_order_model = OrderCreator.create_order_model
+        def new_create_order_model(*args, **kwargs):
+            time.sleep(0.5)
+            return org_create_order_model(creator, *args, **kwargs)
+        creator.create_order_model = new_create_order_model
+
+        # Start 5 threads to place an order concurrently
+        def worker():
+            order_number = threading.current_thread().name
+
+            basket = factories.BasketFactory()
+            basket.add_product(product)
+            place_order(
+                creator, basket=basket, order_number=order_number, user=user)
+
+        exceptions = run_concurrently(worker, num_threads=5)
+
+        assert all(isinstance(x, ValueError) for x in exceptions), exceptions
+        assert len(exceptions) == 0
+        assert Order.objects.count() == 5
+
+        stockrecord = product.stockrecords.first()
+        assert stockrecord.num_allocated == 5
+
+    def test_voucher_single_usage(self):
+        user = AnonymousUser()
+        creator = OrderCreator()
+        product = factories.ProductFactory()
+        voucher = factories.VoucherFactory(usage=Voucher.SINGLE_USE)
+        voucher.offers.add(factories.create_offer(offer_type='Voucher'))
+
+        # Make the order creator a bit more slow too reliable trigger
+        # concurrency issues
+        org_create_order_model = OrderCreator.create_order_model
+        def new_create_order_model(*args, **kwargs):
+            time.sleep(0.5)
+            return org_create_order_model(creator, *args, **kwargs)
```

Comment:
See the comments around it, The idea is to add a delay so that we can consistently trigger the race condition.